### PR TITLE
Number Formatting - Merge Candidate!

### PR DESCRIPTION
**Pull request by er1c**
mini-check point, got all of the numeric tests passing again, still need to cleanup the actual algorithm, add exponent support & add tests for decimals & exponents

### CODE REVIEW

**Comment by cquiroz** [comment]
It starting to look great 👍

**Comment by er1c** [comment]
This implements Integers, Decimals, and Exponents - with some arbitrary pattern support

File: core/src/main/scala/java/text/DecimalFormat.scala [hunk 1]
```diff
@@ -1,57 +1,521 @@
 package java.text
 
-class DecimalFormat(private[this] val pattern: String, private[this] var symbols: DecimalFormatSymbols) extends Format {
+import java.math.{RoundingMode, BigDecimal => JavaBigDecimal, BigInteger => JavaBigInteger}
+import java.util.{Currency, Locale}
+import locales.{DecimalFormatUtil, LocaleRegistry, ParsedPattern}
+import scala.math.{max, min}
+
+// The constructor needs a non-localized pattern
+class DecimalFormat(private[this] val pattern: String, private[this] var symbols: DecimalFormatSymbols)
```
Comment (cquiroz) [inline core/src/main/scala/java/text/DecimalFormat.scala]: is this constructor public on the java api

File: core/src/main/scala/java/text/DecimalFormat.scala [hunk 2]
```diff
@@ -1,57 +1,521 @@
...
+    extends NumberFormat {
+
   def this(pattern: String) = this(pattern, DecimalFormatSymbols.getInstance())
 
-  //def this() = this("???", DecimalFormatSymbols.getInstance())
+  def this() = {
+    this(
+      LocaleRegistry.ldml(Locale.getDefault).flatMap{ _.numberPatterns.decimalPattern }.getOrElse(???),
```
Comment (cquiroz) [inline core/src/main/scala/java/text/DecimalFormat.scala]: The `getOrElse(???)` doesn't look appropriate. It should never happen as `Locale.getDefault` should always return something. Still, I'd rather put a basic default pattern on the `getOrElse`

File: core/src/main/scala/java/text/DecimalFormat.scala [hunk 3]
```diff
@@ -1,57 +1,521 @@
...
+      }
+    }
+
+    // We have the integer portion ready, append it to the builder
+    toAppendTo.append(integerStrBuilder.result.reverse)
+
+    // Compare the target number to the strictly-integer value...
+    targetNumber.compareTo(integerPart) match {
+      // We have a fractional value...
+      case 1  =>
```
Comment (cquiroz) [inline core/src/main/scala/java/text/DecimalFormat.scala]: Perhaps we use a constant that gives more meaning to the `1`?

File: core/src/main/scala/java/text/DecimalFormat.scala [hunk 4]
```diff
@@ -1,57 +1,521 @@
...
+        val unscaledString = fractionStrBuilder.result
+
+        // Drop extra zero's at the end, then add significant '0's to end, then reverse it
+        // 0321 => 321 => 3210 => 0123
+        val fractionStr: String = {
+          val truncatedStr: String = unscaledString.dropWhile(_ == symbols.getZeroDigit)
+
+          truncatedStr +
+          (0 until (fractionMaxDigits - unscaledString.length)).map { _ =>
+            symbols.getZeroDigit
```
Comment (cquiroz) [inline core/src/main/scala/java/text/DecimalFormat.scala]: This bit is probably wrong `getZeroDigit` referns to the pattern for 0 on the date format. It happens to be the char `0` so it mostly works. We should just use the value `0`, or even better the localized number `0`, However the latter is not available yet

File: core/src/main/scala/java/text/FieldPosition.scala [hunk 5]
```diff
@@ -1,5 +1,8 @@
 package java.text
 
+// Make this a super class so we can match against it
+object IgnoreFieldPosition extends FieldPosition(0)
```
Comment (cquiroz) [inline core/src/main/scala/java/text/FieldPosition.scala]: Is this on the java API? If not we should make it package private

File: testSuite/shared/src/test/scala/testsuite/javalib/text/DecimalFormatTest.scala [hunk 6]
```diff
@@ -1,14 +1,663 @@
...
+    assertEquals(-1, f.getGroupingSize)
+
+    f.setDecimalSeparatorAlwaysShown(false)
+    assertFalse(f.isDecimalSeparatorAlwaysShown)
+
+    f.setParseBigDecimal(true)
+    assertTrue(f.isParseBigDecimal)
+  }
+
+  /*
```
Comment (cquiroz) [inline testSuite/shared/src/test/scala/testsuite/javalib/text/DecimalFormatTest.scala]: Why is this one commented out?

File: testSuite/shared/src/test/scala/testsuite/javalib/text/DecimalFormatTest.scala [hunk 7]
```diff
@@ -1,14 +1,663 @@
...
+    assertEquals("123.45670", "123.45670", df.format(123.4567))
+    assertEquals("123.00010", "123.00010", df.format(123.0001))
+    assertEquals("0.0", "0.00000", df.format(0.0))
+    assertEquals("-123", "-123.00000", df.format(-123))
+    assertEquals("-123.45670", "-123.45670", df.format(-123.4567))
+    assertEquals("-123.00010", "-123.00010", df.format(-123.0001))
+    assertEquals("1,000.12340", "1,000.12340", df.format(1000.1234))
+    assertEquals("10,000.12340", "10,000.12340", df.format(10000.1234))
+    assertEquals("10,000,000.12340", "10,000,000.12340", df.format(10000000.1234))
+    //    assertEquals("21,4748,3647", "21,4748,3647", df.format(Int.MaxValue))
```
Comment (cquiroz) [inline testSuite/shared/src/test/scala/testsuite/javalib/text/DecimalFormatTest.scala]: Do you think your new algorithm can handle these cases, where the `MaxValue * multpiler` overflows?

File: core/src/main/scala/java/text/DecimalFormat.scala (hunk 1 above)
Comment (er1c) [inline core/src/main/scala/java/text/DecimalFormat.scala]: Yup, there's 3:
```
DecimalFormat()
Creates a DecimalFormat using the default pattern and symbols for the default locale.
DecimalFormat(String pattern)
Creates a DecimalFormat using the given pattern and the symbols for the default locale.
DecimalFormat(String pattern, DecimalFormatSymbols symbols)
Creates a DecimalFormat using the given pattern and symbols.
```

File: core/src/main/scala/java/text/DecimalFormat.scala (hunk 3 above)
Comment (er1c) [inline core/src/main/scala/java/text/DecimalFormat.scala]: Yeah this whole pattern probably should be cleaned up a little bit, I hate dealing with "compareTo"

File: testSuite/shared/src/test/scala/testsuite/javalib/text/DecimalFormatTest.scala (hunk 7 above)
Comment (er1c) [inline testSuite/shared/src/test/scala/testsuite/javalib/text/DecimalFormatTest.scala]: probably

File: core/src/main/scala/java/text/DecimalFormat.scala (hunk 2 above)
Comment (cquiroz) [inline core/src/main/scala/java/text/DecimalFormat.scala]: Right, it shouldn't be called ever but I'm not too happy about throwing an exception sincethe JVM never does that. We could extract the default pattern in the JVM with locale english and hardcode it there

Comment (er1c) [inline core/src/main/scala/java/text/DecimalFormat.scala]: Yeah, unless you thought were getting the localized one, but you forgot to do something else :) - I might crash early here - `the https://pragprog.com/the-pragmatic-programmer/extracts/tips` but in terms of how simple this is, I'm sure the getDefault will throw an error if they weren't installed properly/etc. so I think I'm fine with the default pattern approach

File: testSuite/shared/src/test/scala/testsuite/javalib/text/DecimalFormatTest.scala (hunk 7 above)
Comment (er1c) [inline testSuite/shared/src/test/scala/testsuite/javalib/text/DecimalFormatTest.scala]: that specific test is testing the max fractions, so setting the max value is pointless, but I'll make sure the multiplier test has them

**Comment by cquiroz** [comment]
This seems very much ready for release. @er1c Do you want to keep updating the PR or would you be ready for a release? Could you rebase onto master to get the latest updates

**Comment by er1c** [comment]
@cquiroz I just rebased and pushed, hopefully tests will pass fine, for the moment I'm done making changes

**Pull request merged**